tests/test_mutations.py

```diff
@@ -472,47 +705,288 @@ def test_keep_mutation_parent_zero_rate(self):
+            self.sites.append(site_item)
+
+    def populate_tables(self, tables):
+        for site in self.sites:
+            site_id = tables.sites.add_row(
+                    site.position,
+                    site.ancestral_state,
+                    site.metadata)
+            for mutation in site.mutations:
+                tables.mutations.add_row(
+                            site_id, mutation.node, mutation.derived_state,
+                            parent=mutation.parent, metadata=mutation.metadata)
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.rate[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_positions = [s.position for s in self.sites]
+                    index = np.searchsorted(site_positions, position)
+                    if (index == len(self.sites)
+                       or site_positions[index] != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele(rng)]
+                        site = SiteItem(
+                                position=position,
+                                ancestral_state=ancestral_state,
+                                metadata=b'',
+                                mutations=[])
+                        self.sites.insert(index, site)
+                    else:
+                        site = self.sites[index]
+                    # keep mutations sorted in decreasing order by time
```

Comment:
Is there any point in keeping the mutations sorted? Why not do this at the end when all the mutations are known? If there's a large number of mutations and we need to keep them sorted at all times, then we should be using an AVL tree for this. If not, then a single sort at the end is much simpler and more efficient.

Reply:
Hm, ok: once we have a `time` attribute then we can do this. In the meantime we do need to keep the `kept` mutations in the same order they came in, in case there are more than one on the same branch. (And: I'm imagining not optimizing for the 'large number of mutations' case, so a sort after would be more efficient.)

Reply:
Well, I guess I can just stabilize qsort for the time being.

Reply:
Ah, ok. How about sorting on ``(-time, parent)`` then? Assuming the incoming mutations are well formed, this should work as tskit requires that any mutation is listed before its parent.

Reply:
oh, good call.